field/TinyLang/Field/Typed/TypeChecker.hs

```diff
@@ -220,25 +220,32 @@ checkExpr m = do
     let uniMismatch = typeMismatch tM uni mUni
     withGeqUniM uni mUni uniMismatch tM
 
+-- TODO:  Figure how to optimise
 checkProgram
     :: forall m f. (MonadTypeChecker m, TextField f)
     => R.Program R.Var f -> m (T.Program f)
-checkProgram = traverse checkStatement
+checkProgram (R.Program (R.Statements stmts)) =
+    T.Program . T.Statements <$> foldMapA checkStatement stmts
 
 
 {-| Type checking judgement for statements of form
 -}
 checkStatement
     :: forall m f. (MonadTypeChecker m, TextField f)
-    => R.Statement R.Var f -> m (T.Statement f)
+    => R.Statement R.Var f -> m [T.Statement f]
 checkStatement (R.ELet var m) = do
     Some uniVar@(T.UniVar uni _) <- inferUniVar var
-    T.withKnownUni uni $ T.ELet uniVar <$> checkExpr m
+    T.withKnownUni uni $ (:[]) . T.ELet uniVar <$> checkExpr m
 checkStatement (R.EAssert m) =
-    T.EAssert <$> checkExpr m
-checkStatement (R.EFor var start end stmts) = do
-    tVar <- makeVar $ R.unVar var
-    T.EFor (T.UniVar Field tVar) start end <$> traverse checkStatement stmts
+    (:[]) . T.EAssert <$> checkExpr m
+-- TODO:  Inject var into environment of stmts
+checkStatement (R.EFor var start end stmts) =
+    foldMapA iter [start .. end]
+    where iter i = do
+              tVar <- makeVar $ R.unVar var
+              let uVar = T.UniVar Field tVar
+              tStmts <- foldMapA checkStatement stmts
+              pure $ T.ELet uVar (fromIntegral i) : tStmts
```

Comment:
It should be `EConst $ fromIntegral i` instead of `fromIntegral i` (provided the former type checks). Otherwise we end up with stuff like `let j_10 = 1 + (1 + 1);` in test cases below.

Reply:
Done in 115ec82